### PR TITLE
Verbitim lemmatizer

### PR DESCRIPTION
**Pull request by adam-benjellounSG**
# PR related to https://github.com/MAIF/melusine/issues/127#issue-1138862604

Adding Lemmatizer object which wraps Spacy and Lefff lemmatizers.
Fully compatible with TransformerScheduler as well as sklearn pipelines.
Enables the user to provide name of input and output columns.

Test plan (required)

```
from melusine.nlp_tools.lemmatizer import Lemmatizer

spacy_lemma_sm = Lemmatizer('clean_body', 'lemma_spacy_sm', engine='spacy', engine_conf='fr_core_news_sm')
spacy_lemma_sm.fit_transform(df_email)
```

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
Implementation of wrapping around spacy and lefff lemmatizers compatible with TransformerScheduler architecture. The Lemmatizer object is compatible with sklearn pipelines and is built around an sklearn Transformer.

## Type of change

Please delete options that are not relevant.

- [X ] New feature (non-breaking change which adds functionality or feature that would cause existing functionality to not work as expected)
- [ X] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
conda create -n melusine-verbitim python=3.6
conda activate melusine-verbitim
pip install melusine[lemmatizer]

- [ ] Create a Lemmatizer object of each engine
```
from melusine.nlp_tools.lemmatizer import Lemmatizer
spacy_lemma_sm = Lemmatizer('clean_body', 'lemma_spacy_sm', engine='spacy', engine_conf='fr_core_news_sm')
lefff_lemma = Lemmatizer('clean_body', 'lemma_lefff', engine='Lefff', engine_conf='fr_core_news_sm')
```
- [ ] Insert Lemmatizer object in a pipeline with TransformerScheduler objects
```
from sklearn.pipeline import Pipeline
from melusine.utils.transformer_scheduler import TransformerScheduler
from melusine.prepare_email.manage_transfer_reply import check_mail_begin_by_transfer, update_info_for_transfer_mail, add_boolean_transfer, add_boolean_answer
from melusine.prepare_email.build_historic import build_historic
from melusine.prepare_email.mail_segmenting import structure_email
from melusine.prepare_email.body_header_extraction import extract_last_body
from melusine.prepare_email.cleaning import clean_body

ManageTransferReply = TransformerScheduler(
functions_scheduler=[
    (check_mail_begin_by_transfer, None, ['is_begin_by_transfer']),
    (update_info_for_transfer_mail, None, None),
    (add_boolean_answer, None, ['is_answer']),
    (add_boolean_transfer, None, ['is_transfer'])
])

EmailSegmenting = TransformerScheduler(
functions_scheduler=[
    (build_historic, None, ['structured_historic']),
    (structure_email, None, ['structured_body'])
])

Cleaning = TransformerScheduler(
functions_scheduler=[
    (extract_last_body, None, ['last_body']),
    (clean_body, None, ['clean_body'])
])

from melusine.nlp_tools.lemmatizer import Lemmatizer

spacy_lemma_sm = Lemmatizer('clean_body', 'lemma_spacy_sm', engine='spacy', engine_conf='fr_core_news_sm')
lefff_lemma = Lemmatizer('clean_body', 'lemma_lefff', engine='Lefff', engine_conf='fr_core_news_sm')

prepare_data_pipeline = Pipeline([
  ('ManageTransferReply', ManageTransferReply),
  ('EmailSegmenting', EmailSegmenting),
  ('Cleaning', Cleaning),
  ('spacy_lemma_sm', spacy_lemma_sm),
  ('lefff_lemma', lefff_lemma)
])

df_email = prepare_data_pipeline.fit_transform(df_email)
```

**Test Configuration**:
* OS:Windows 10
* Python version:3.6.13
* Melusine version: 2.3.2

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: melusine/nlp_tools/lemmatizer.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import numpy
+import spacy
+from spacy.language import Language
+from spacy_lefff import LefffLemmatizer, POSTagger, Downloader
+
+from melusine.utils.verbitim_transformer import VerbitimTransformer
+
+
+@Language.factory('melt_tagger')
+def create_melt_tagger(nlp, name):
```
Comment (TFA-MAIF) [inline melusine/nlp_tools/lemmatizer.py:14]: Hello, could you add docstring to understand what the following functions do?
Best,

File: melusine/nlp_tools/lemmatizer.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                 in_col: str,
+                 out_col: str,
+                 engine: str,
+                 engine_conf: Union[dict, str, None],
+                 n_jobs: int = 1,
+                 batch_size: int = 1000
+                 ):
+        """
+        Initialize a Socle NLP Lemmatizer object
+        :param in_col:  input column of dataframe.
```
Comment (TFA-MAIF) [inline melusine/nlp_tools/lemmatizer.py:50]: Could you use numpy format for docstring? here is the link: https://numpydoc.readthedocs.io/en/latest/format.html

File: melusine/nlp_tools/lemmatizer.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        """
+
+        self.in_col = in_col
+        self.out_col = out_col
+        self.engine = engine
+        self.engine_conf = engine_conf
+        self.n_jobs = n_jobs
+        self.batch_size = batch_size
+
+        self.logger = logging.getLogger(__name__)
```
Comment (TFA-MAIF) [inline melusine/nlp_tools/lemmatizer.py:67]: We do not recommand adding logger into the object but preferably at a module level.

File: melusine/utils/verbitim_transformer.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class VerbitimTransformer(BaseEstimator, TransformerMixin):
+
```
Comment (TFA-MAIF) [inline melusine/utils/verbitim_transformer.py:5]: Could you add docstring?

File: melusine/nlp_tools/lemmatizer.py (hunk 1 above)
Comment (TFA-MAIF) [reply melusine/nlp_tools/lemmatizer.py:14]: + type hinting for every function with format compatible for Python <3.9

def ma_fonction(malist: List[int], mastring: str) -> Union[List, str]:

**Review by TFA-MAIF — CHANGES_REQUESTED** [review]
Review done by Hugo, Charles and Tiphaine

**Comment by adam-benjellounSG** [comment]
Hello, 
We performed changes, please let us know if there are any points we may have missed.
Best Regards,
Adam

**Review by TFA-MAIF — APPROVED** [review]

**Pull request merged**